planner/types.go

```diff
@@ -213,6 +216,84 @@ func (e *edges) ClearSuccessors() {
 	e.successors = e.successors[0:0]
 }
 
+func (e *edges) ClearPredecessors() {
+	e.predecessors = e.predecessors[0:0]
+}
+
+func (e *edges) shallowCopy() edges {
+	newEdges := new(edges)
+	copy(newEdges.predecessors, e.predecessors)
+	copy(newEdges.successors, e.successors)
+	return *newEdges
+}
+
+// MergeLogicalPlanNodes merges top and bottom plan nodes into a new plan node, with the
+// given procedure spec.  The returned node will have its predecessors set to the predecessors
+// of "bottom", however, it's successors will not be set---it will be the responsibility of
+// the planner to attach the merged node to its successors.
+func MergeLogicalPlanNodes(top, bottom PlanNode, procSpec ProcedureSpec) PlanNode {
```

Comment:
Can you put a visual description of this one like you did for `SwapPlanNodes` showing that it's a linear sequence of nodes? Those are really helpful, I think.